metals/src/main/scala/scala/meta/internal/metals/debug/EndpointLogger.scala

```diff
@@ -0,0 +1,72 @@
+package scala.meta.internal.metals.debug
+
+import java.io.PrintWriter
+import java.time.Clock
+import java.time.format.DateTimeFormatter
+import java.util.Collections
+import java.util.function.Consumer
+import com.google.gson.GsonBuilder
+import org.eclipse.lsp4j.jsonrpc.MessageConsumer
+import org.eclipse.lsp4j.jsonrpc.debug.json.DebugMessageJsonHandler
+import org.eclipse.lsp4j.jsonrpc.json.MessageJsonHandler
+import org.eclipse.lsp4j.jsonrpc.messages.Message
+import org.eclipse.lsp4j.jsonrpc.messages.NotificationMessage
+import org.eclipse.lsp4j.jsonrpc.messages.RequestMessage
+import org.eclipse.lsp4j.jsonrpc.messages.ResponseMessage
+import scala.meta.internal.metals.debug.EndpointLogger.Direction
+import scala.meta.internal.metals.debug.EndpointLogger.Received
+import scala.meta.internal.metals.debug.EndpointLogger.Sent
+import scala.meta.internal.metals.debug.EndpointLogger.time
+
+final class EndpointLogger(endpoint: RemoteEndpoint, logger: PrintWriter)
+    extends RemoteEndpoint {
+  private val writer: MessageJsonHandler = EndpointLogger.serializer
+
+  override def consume(message: Message): Unit = {
+    try log(Sent, message)
+    finally endpoint.consume(message)
+  }
+
+  override def listen(messageConsumer: MessageConsumer): Unit = {
+    endpoint.listen { message =>
+      try log(Received, message)
+      finally messageConsumer.consume(message)
+    }
+  }
+
+  override def cancel(): Unit = endpoint.cancel()
+
+  private def log(direction: Direction, message: Message): Unit = {
+    logger.println(s"[Trace][$time] $direction ${typeOf(message)}:")
```

Comment:
Does it work with the inspector tool? It would be nice to keep it compatible.

Reply:
Sadly, no - the format is too much of a hassle to implement it manually. Reusing the implementation present in the `lsp4j.jsonrpc` library is also not viable because it is tightly coupled with their launcher implementation.

At one point we might want to change the format to accomodate the inspector but let's delay it for now as it is better to have some logs than none at all.